nodejs/lib/util/webstorage.js

```diff
@@ -0,0 +1,109 @@
+"use strict";
+var WebStorage = (function() {
+	var path = require('path');
+	var os = require('os');
+	var mkdirs = require('node-mkdirs');
+	var JSONStorage = require('node-localstorage').JSONStorage;
+
+	function storedItem(value, ttl) {
+		var item = {
+			value: value
+		};
+		if(ttl) {
+			item.expires = (Date.now() + ttl);
+		}
+		return item;
+	}
+
+	function valid(item) {
+		var expires = item.expires;
+		return !expires || (Date.now() < expires);
+	}
+
+	function WebStorage(client) {
+		this.client = client;
+		this.storagePath = null;
+		this.localStorage = null;
+		this.sessionStorage = new Map();
+	}
+
+	WebStorage.prototype.supportsLocal = true;
+	WebStorage.prototype.supportsSession = true;
+
+	WebStorage.prototype.initPath = function() {
+		var storagePath = this.storagePath;
+		if(!storagePath) {
+			var options = this.client.options, appId;
+			if(options.key) {
+				appId = options.key.split('.')[0];
+			} else {
+				var authOptions = this.client.auth.authOptions,
+					token = authOptions.tokenDetails.token;
+				if(!token) {
+					throw new Error('Unable to initialise WebStorage; no key or token available');
```

Comment:
Aside from comment above about getting the app ID, if it can't initialize storage for some reason, it should just set `supportsLocal` and `supportsSession` to `false` and return, not throw an exception and abort the whole client init.

Reply:
> or better, keep the exception here, but in init wrap initPath() and the JSONStorage init in a try/catch, and set the supports vars to false in the catch

Yes